1.18: Revamp & Extension of the PermissionAPI

JTK222 opened the pull request:
The PermissionAPI is a little used feature of Forge, this PR tries to make it more versatile and therefore increase it's usage.
Following changes are and will be made to achieve this:
| Finished |   Task     | Reason |
| ------------- | ------------- | ----- |
| ✓ | Add a PermissionType class to allow more than just boolean based permissions | It can often be necessary to check for more than just a boolean value. A good example for this would be to check how many chunks a user can load. |
| ✓ | Add better DefaultHandler | The current default handler isn't granular enough for many people, even vanilla provides more granularity |
| ✓ | Replace String permission with PermissionNodes | This change was made in favor of the above 2 changes |
| ✓ | Remove IContext | The Context is ambiguous in all cases I could think of, all required context for a permission Check should be obtainable from the Entity for which the permissions are checked |
| ✓ | Add DynamicNodeContext | This is supposed to be a less ambiguous replacement for the IContext, this will also make it possible to enforce PermissionNode registration as PermissionNodes will not have to be created dynamically. (For example in case of worldguard, rathern then creating a new node for each region: worldguard.canbypassmembers.\<regionname>, the region name would be passed in as a DynamicNodeContext) |
| ✓ | Add a Registry Event for PermissionHandlers | This opens the possibility to sort out compatibility issues if multiple mods implement a PermissionHandler, users could choose which one to use trough a config option |
| ✓ | Add a registry event for PermissionNodes | Registering permission nodes does for example allow auto completion in commands, dumping a list of available permissions and other things which a Permission mod might come up with. Additionally it might allow for better optimizations. Registering PermissionNodes might become necessary, as dynamic permissions can be handled trough the DynamicNodeContext |

This is a breaking change and should not be merged before 1.17

Here's a mockup of how I want the usage to look:
```java
public class Mod{
   public static PermissionNode<Integer> LOADED_CHUNKS = new PermissionNode<>("modid.chunk_loading.maximum", PermissionTypes.INT, player -> 8); //Could also return value based on OP level
      public static PermissionNode<Boolean> CAN_BYPASS_MEMBERS = new PermissionNode<>("modid.regions.can_bypass", PermissionTypes.BOOL, player -> player.getOPLevel() > 3); //the OP check is a bit abstracted here for simplicity
   
   public static void registerNodes(PermissionNodeGatherEvent event){
      event.addNodes(LOADED_CHUNKS, CAN_BYPASS_MEMBERS);
   }

   //Usage
   int chunkLoadingMax = PermissionAPI.getPermission(player, LOADED_CHUNKS);

   boolean canBypassMembers = PermissionAPI.getPermission(player, CAN_BYPASS_MEMBERS, DynamicContext.Builder().add("region", "spawn"));
}
```

JTK222 commented:
As 1.17 just dropped I'd like to push this and get more potential opinions for this.
I'll try to finish this PR next week therefore having opinions on the design idea beforehand would be nice.
(I want this to be an API that more people are happy with than the previous one, so please if there is anything you don't like about the idea/example above, post it here)

Alpvax commented:
RE: removing IContext: I agree that the previous approach was too ambiguous, but I could see an argument for an object containing player and opLevel (because I imagine the default implementation for virtually all nodes will want to retrieve that).

What are the opinions on parent nodes (e.g. a worldguard.canbypassmembers catch-all or default node)?
We could adjust the node creation to me much more like the way commands are registered, with a chainable (type, default) function.
Alternatively we could have a default-redirect (i.e. if this permission is not explicitly specified, redirect to querying permission x).

Of course all of that could be left to the permission managing mods themselves, it's just whether we want to give the mods the control (at the risk of making it more complicated) or let the managing mods do the work (with the potential for nodes not quite working the way the mod intended).

TheSilkMiner commented:
IMO there's no reason to have the concept of a parent node. This is a general API so it should be able to work regardless of the implementation: the decision on how to behave in case of hierarchy should be left to the permission handler, since a mod should always query with the correct and most specific permission key available for the action it wants to perform. 
Speaking of hierarchy, I don't believe there should be any assumption or enforcement of this concept. Registry entries in vanilla are, in fact, not hierarchical in nature: `mod:thing/stuff` is at the same level as `mod:thing` itself. I don't believe permissions should be special cased in this way, which also adds burden onto the permission handler maintainer and mod maker, that now have to think about the implications of where their nodes are located.
Another reason for not having a hierarchy is the possibly non-sensical results that may occur. Take for instance a mod that allows chunk claiming, which has the following permissions: `cc:chunks/claim`, `cc:chunks/bypass`, `cc:chunks/limit`. The first two are boolean permissions and the last one is an integer permission. If I were now given the `cc:chunks` permission, what does that mean? The two booleans could be set to `true`, but what about the integer one? What about something more esoteric like a text component?

Alpvax commented:
Yes, that seems a sensible approach.

> Take for instance a mod that allows chunk claiming, which has the following permissions: `cc:chunks/claim`, `cc:chunks/bypass`, `cc:chunks/limit`. The first two are boolean permissions and the last one is an integer permission. If I were now given the `cc:chunks` permission, what does that mean? The two booleans could be set to `true`, but what about the integer one? What about something more esoteric like a text component?

For reference, I meant for the dynamic nodes (i.e. a fallback for when there is no specific node registered). Regarding the different types, that is exactly why I was thinking it might be worth building it into the API and giving the mod control, rather than the manager trying to handle it in a general way.

JTK222 commented:
> (i.e. a fallback for when there is no specific node registered).

In best case unregistered nodes won't be a thing, that's what the dynamic context (as shown in the example)
Is for, how permission handlers handle that, or if they handle those at all is up to them.

Alpvax commented:
> In best case unregistered nodes won't be a thing, that's what the dynamic context (as shown in the example)
> Is for, how permission handlers handle that, or if they handle those at all is up to them.

My meaning when I said "registered" was not in the sense of a registered permission node, but a registered permission for that specific node, sorry for the confusion.
I will attempt to explain myself better with an example:

> ModA registers a dynamic permission node: "moda.candosomething.[dynamic playername]". Granting this permission allows the player to do something with the specified [dynamic playername].
> 
> Permission manager ModB recieves a query for "moda.candosomething.Alpvax". It doesn't have a specified rule for [Alpvax], so it should fall back to a default (which I'm suggesting is the "parent" node: "moda.candosomething").

Currently I envisage this as having to register a "moda.candosomething" node as well as a "moda.candosomething.[dynamic]" node. I may be misunderstanding the registration of the dynamic part, but in your example it doesn't show how the manager would query against the dynamic context. Hence my suggestion of a chainable builder system.

Another scenario is when a mod registers serveral increasingly specific nodes (as some bukkit plugins I used to use did) e.g.:

> ModA registers "moda.cando.something", "moda.cando.something.specific", "moda.cando.something.else"...
> ModB recieves a request for "moda.cando.something.specific", but it only has a rule defined for "moda.cando.something" (as a general "Player can do everything (unless otherwise specified in a more specific rule)".

If there was a way for ModA to say "if this rule isn't specified, defer to rule 'XXX'" then ModB would just be able to say "this rule isn't defined, query 'XXX'" (or it could cache the rules on load into a single lookup table or some other implementation). It would put the control into the hands of the people writing the rules, rather than the manager making some assumption that rules for "moda.something" must apply to "moda.something.*".

The alternative is to say that managers _should_ error on missing permissions, but then users might have to look through loads of default permissions in a file to find the one they need to change, and they may need to change the value in multiple places (e.g. a quantity with 3 specialisations might need to be changed in 3 places).

JTK222 commented:
Having the dynamic part defined as part of a registered node might indeed be a good Idea.
So something like this:
```java
public static PermissionNode<Boolean> CAN_BYPASS_MEMBERS = new PermissionNode<>("modid.regions.can_bypass.%region%", PermissionTypes.BOOL, player -> false);
boolean canBypassMembers = PermissionAPI.getPermission(player, CAN_BYPASS_MEMBERS, DynamicContext.Builder().add("region", "spawn"));
```
However, I'd prefer not to have it a part. I kinda imagine it in a block property <-> state way.
So you'd have `modid.regions.can_bypass[region=<region>]`
But it might be a good Idea to enforce registration of those "dynamic" properties. (not the values, only the fields)

Alpvax commented:
> But it might be a good Idea to enforce registration of those "dynamic" properties. (not the values, only the fields)

We could always make it register a class, then people could use enums or e.g. a `Region` Object (or just string/ integer etc.)

TheSilkMiner commented:
As for the specificity you were mentioning before, a mod can already do that as it can defer the check again to the permission handler, if such behavior were desired. You'd just need to expand the default function to also get a context passed in.

For the dynamic portion, I am definitely in favor of registering the **required** dynamic parts (I am thinking that maybe a permission handler may add some stuff into the context for its own book-keeping or logging or something, and those shouldn't need to be validated). You could have something like (pardon my Kotlin) 

```kotlin
val region = PermissionContext.registerDynamic("region", Region::class) // Region implements StringSerializable or however that class is called

val node = PermissionNode("modid.regions.bypass", PermissionType.BOOL, arrayOf(region)) { player, context -> true }

val check = PermissionHandler.hasPermission(node, player, mapOf(region to Region["overworld"]))
```

Alpvax commented:
> As for the specificity you were mentioning before, a mod can already do that as it can defer the check again to the permission handler, if such behavior were desired.

I'm not sure I like that as much. In that case the permission manager has to return a special "not specified" value (or all permissions have to become Optionals). Or did you mean that the mod could do that in the default handler? In that case it is necessary for the manager mod to cache the permissions. ~~Imagine some permissions mod which retrieves the permissions from an external server, or loads them from file on each request: the operation could then potentially take some time, increasing with each lookup, rather than just looking up the correct permission to use and looking up the value for a single permission.~~
Edit: I'm not sure that last statement is true, as the manager would still have to look up whether or not a rule is specified for the requested permission.

---

> `PermissionContext.registerDynamic("region", Region::class) // Region implements StringSerializable`

We might want to design an interface which can save more than just strings. There is no guarantee that the manager mod wants to save the permissions as strings.
We could have an `IDynamicContext` interface with from/to an `ISerialAdapter` (or some other name) interface, which would have methods similar to the config object or packetbuffer (i.e. save/load strings, primitives, lists etc.). That way, manager mods could use whatever format they want to (JSON, TOML, NBT, custom binary format, database, etc.)

TheSilkMiner commented:
How so? Consider the following permission:

```kotlin
val interactNode = PermissionNode("stuff/interact", PermissionType.BOOL, arrayOf()) { player, context -> true }
val specificInteractNode = PermissionNode("stuff/interact/specific", PermissionType.BOOL, arrayOf()) { player, context ->
    PermissionHandler.hasPermission(interactNode, player, context)
}
```

This default handler simply queries the parent permission. The advantage of this is that now the Mod can decide whether stuff is in a parent/child relationship without the permission handler having to care or the API needing to impose a hierarchy that mods may not want to use.

JTK222 commented:
I think I understand the issue you want to bring up.
In a case with 2 permissions `perm.a` and `perm.a.b` it should not go up the hierarchy tree.
If a user doesn't choose to set a perm with a Permission Handler mod, you can assume that they want to use whatever is the default value. And each mod that adds perms, can decide that default on it's own.

> We might want to design an interface which can save more than just strings. There is no guarantee that the manager mod wants to save the permissions as strings.

And I am strongly against that, that's exactly what the old IContext did, and there is no sense in removing it just to add it with a new name. The dynamic context is there to cover a few edge cases. If you need anything else than a string or something that can not be converted to a string, you're using the API incorrectly.
I am already skeptical to allow usage of classes vs simple Strings, but I can see that it makes the usage a bit easier.
And it would allow a Permission Handler to add mod specific optimization.
(But mods should **never** expect that a permission handler uses anything else than the strings of Dynamic Context)

Alpvax commented:
Yes. That all sounds good to me. I wasn't trying to persuade you, just bouncing some ideas around. I don't want it to be left in a position where it doesn't handle the use cases of many mods and then ends up not getting used.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=7780) <br/>All committers have signed the CLA.

JTK222 commented:
Current Example mod so people can see how it works more or less.
```java
   private static final PermissionNode<Boolean> blob = new PermissionNode<>("permissiontest", "test.blob", PermissionTypes.BOOL, (player, context) -> true);

   @SubscribeEvent
   public void registerNodes(PermissionGatherEvent.Nodes event)
   {
      event.addNodes(blob);
   }

   @SubscribeEvent
   public void registerCommands(RegisterCommandsEvent event)
   {
      event.getDispatcher().register(Commands.literal("permissiontest")
            .requires(src -> src.getEntity() != null && src.getEntity() instanceof ServerPlayer && PermissionAPI.getPermissions((ServerPlayer) src.getEntity(), blob))
            .executes(context -> {
               context.getSource().sendSuccess(new TextComponent("Blob"), false);

               return 1;
            }));
   }
```
Currently only missing thing is documentation, and potentially a more complex test mod (if required)

Small Explanation: PermissionDynamics need to be registered together with a PermissionNode.
They are intended to be handled like block properties, and can be added optionally when checking permissions.
However adding more properties than registered should throw an error.

P.s. The Jenkins ran into a race condition with another PR, it should work tomorrow when I make the PR with the docs

marchermans commented on src/main/java/net/minecraftforge/server/permission/PermissionAPI.java:
```diff
@@ -19,80 +19,73 @@
-        return permissionHandler;
-    }
+   public static Collection<PermissionNode<?>> getRegisteredNodes()
+   {
```
Bracket on the wrong line.

JTK222 replied:
Formatting was correct after all. Resolved on the Discord Server.
<https://canary.discord.com/channels/313125603924639766/852298000042164244/883245371013337148>

marchermans commented on src/main/java/net/minecraftforge/server/permission/PermissionAPI.java:
```diff
@@ -19,80 +19,73 @@
+    * PermissionNodes default handler.
+    */
+   public static <T> T getPermissions(ServerPlayer player, PermissionNode<T> node, PermissionDynamicContext<? extends StringRepresentable>... context)
+   {
```
Again

marchermans commented on src/main/java/net/minecraftforge/server/permission/PermissionAPI.java:
```diff
@@ -19,80 +19,73 @@
+    * and replaced the current PermissionHandler with the one returned by the event.</p>
+    */
+   public static void gatherPermissionHandler()
+   {
```
Generally check your formatting.
Won't continue reviewing the formatting for the remaining files.

marchermans commented on src/test/java/net/minecraftforge/debug/PermissionTest.java:
```diff
@@ -0,0 +1,42 @@
+   @SubscribeEvent
+   public void registerCommands(RegisterCommandsEvent event)
+   {
+      event.getDispatcher().register(Commands.literal("permissiontest")
+            .requires(src -> src.getEntity() != null && src.getEntity() instanceof ServerPlayer && PermissionAPI.getPermissions((ServerPlayer) src.getEntity(), blob))
+            .executes(context -> {
+               context.getSource().sendSuccess(new TextComponent("Blob"), false);
+
+               return 1;
+            }));
```
It would be nice if you could extend the examples here.

Additionally besides a "positive" test, a "negative" test as well as examples with value types other then `Boolean` would be a good idea to add and test.

JTK222 replied:
The test mod has been expanded, is it ok now or would you like more?

marchermans reviewed (changes requested):
Generally the implementation and API looks good.
However the formatting is of, as well as the test mod is a bit small for the size of the change.

Please fix the formatting, and expand the test mod.

TheSilkMiner commented on src/main/java/net/minecraftforge/server/permission/PermissionAPI.java:
```diff
@@ -19,80 +19,73 @@
+    * @param node the PermissionNode for which you want to query
+    * @param context optional array of PermissionDynamicContext, single entries will be ignored if they weren't
+    *                registered to the node
+    * @param <T></T> type of the queried PermissionNode
```
Unnecessary closing tag: `<T></T>`. The parameter is fine with `<T>`

TheSilkMiner commented on src/main/java/net/minecraftforge/server/permission/handler/IPermissionHandler.java:
```diff
@@ -0,0 +1,64 @@
+    * @param node the PermissionNode for which you want to query
+    * @param context optional array of PermissionDynamicContext, single entries will be ignored if they weren't
+    *                registered to the node
+    * @param <T></T> type of the queried PermissionNode
```
Again, `<T></T>`

TheSilkMiner commented on src/main/java/net/minecraftforge/server/permission/handler/IPermissionHandler.java:
```diff
@@ -0,0 +1,64 @@
+
+import java.util.Collection;
+
+public interface IPermissionHandler
```
I feel like this class is missing documentation on what a permission handler actually is and how it should behave. It can also be short and simple.
Also, I would personally add an `@implNote` specifying how the Handlers are managed or an `@apiNote` describing how a client should register it. Currently there's nothing that ties the handler to the event, so it might be nice to have it as a sort of footnote.

TheSilkMiner commented on src/main/java/net/minecraftforge/server/permission/events/PermissionGatherEvent.java:
```diff
@@ -0,0 +1,94 @@
+         Collections.addAll(this.nodes, nodes);
+      }
+
+      public void addNodes(Collection<PermissionNode<?>> nodes)
```
Why `Collection` and not `Iterable`?

JTK222 replied:
Could you elaborate what exactly you mean?
I don't see why Iterable should be used here.

TheSilkMiner replied:
I was asking if there was a clear reason as for why requiring a collection instead of `Iterable`s in general. The latter would allow for more possibilities, like Kotlin `Sequence`s, being used.

JTK222 replied:
The only reason was that Collections.addAll requires a collection.
But sure, I can change that.

TheSilkMiner commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionDynamic.java:
```diff
@@ -17,30 +17,20 @@
+import net.minecraft.util.StringRepresentable;
 
-public class PlayerContext extends Context
+/**
+ * <p>PermissionDynamics can be seen like Block Properties.
+ * An example usage might be a dimension inside a check for building permissions:
+ * "modid.build.allowed[dimension=overworld]" </p>
+ * <p>Note: The string representation is just an example and may be different for other PermissionHandlers</p>
+ * @param <T> type of the PermissionDynamic, must implement {@link StringRepresentable}
+ */
+public record PermissionDynamic<T extends StringRepresentable> (Class<T> typeToken, String name)
```
Don't get me wrong, but I don't find the documentation clear.
A better idea would be
```
Represents a key that can be used to build a {@link PermissionDynamicContext}.

<p>Keys, along with their associated values, can be used to provide additional context for a permission handler in determining whether to grant permission for an actor and a specific node.</p>

<p>As an example usage, a dimension context key could be used inside a building permission check to ensure that the actor can build given those constraints.</p>
```

I'd also rename this class to `PermissionDynamicContextKey` to make it easier to understand that it doesn't store a value. Also, it could be nice to indicate that there is no restriction on the values of the value (pun not intended), since that's the exact opposite of how block properties behave.

TheSilkMiner commented on src/main/java/net/minecraftforge/server/permission/events/PermissionGatherEvent.java:
```diff
@@ -0,0 +1,94 @@
+         return nodes;
+      }
+
+      public void addNodes(PermissionNode<?>... nodes)
```
What about two nodes with the same `nodeName`?

JTK222 replied:
Do Hashset work in Buckets? If not they'll be removed. As they have the same hashes.

sciwhiz12 replied:
Maybe it should be considered to explicitly test if the node is already listed in this event, and throw an exception if so?

TheSilkMiner commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionDynamicContext.java:
```diff
@@ -17,33 +17,32 @@
-import javax.annotation.Nullable;
-
-public class WorldContext extends Context
+/**
+ * Pair of a PermissionDynamic and a value of the corresponding type.
+ * Use {@link PermissionDynamic#createContext(StringRepresentable)} for constructing.
+ */
+public class PermissionDynamicContext<T extends StringRepresentable>
```
I'd suggest to also take a look at this documentation after fixing the `PermissionDynamic` one, as there may be a need for some rewording.
I'd personally add that there is no restriction on how a permission handler should represent this context, so it can be in a blockstate format or whatever else.

TheSilkMiner replied:
Why is the class not `final`?

TheSilkMiner commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionNode.java:
```diff
@@ -0,0 +1,121 @@
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.level.ServerPlayer;
+
+/**
+ * @see PermissionNode#PermissionNode(String, PermissionType, PermissionResolver, PermissionDynamic[])
+ * @param <T> PermissionType see {@link PermissionTypes}
+ */
+public class PermissionNode<T>
```
```
Represents the basic unit at the heart of the permission system.

<p>A permission indicates the ability for an actor to perform an action, in its most general sense. In the permission system, all permissions are encoded as instances of this class, optionally integrated by a {@link PermissionDynamicContext}.</p>

<p>A node is uniquely identified by its `nodeName`, which is a dot-separated string providing meaning to the node itself. The suggested structure of the name is `modid.name`, where `modid` is the ID of the owner of the node. This API <strong>does not require</strong> any implicit hierarchy, so `modid.name` is not considered a parent of `modid.name.sub`. Such decisions are left to the {@link IPermissionHandler}.</p>

<p>Each node also has an associated {@link Permission Type}, detailing its exact type, along with a {@link PermissionResolver} that is used to obtain the default value of the permission. More information can be found on their documentation.</p>

<p>A node can also be bound to additional {@link PermissionDynamic}s, which are used when querying the availability of the permission. There are no restrictions on their amount. It is also not mandatory to provide a value for all dynamics in a permission query. See the above link for more information.</p>

<p>Each node should be registered via the {@link Event} and stored statically in a field. That instance should then be reused every-time a permission check needs to be performed via {@link PermissionAPI#check}.</p>
```

Not the best, but I feel this describes more coherently the class and gives more information for someone that is just getting started. A good idea would be to reference nodes in the documentation for `PermissionAPI` too, as to give a clearer connection.

JTK222 replied:
You, are a master at writing Java Docs :eyes: Love it!

TheSilkMiner commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionNode.java:
```diff
@@ -0,0 +1,121 @@
+ * @see PermissionNode#PermissionNode(String, PermissionType, PermissionResolver, PermissionDynamic[])
+ * @param <T> PermissionType see {@link PermissionTypes}
+ */
+public class PermissionNode<T>
```
`final`?

TheSilkMiner commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionNode.java:
```diff
@@ -0,0 +1,121 @@
+   private final String nodeName;
+   private final PermissionType<T> type;
+   private final PermissionResolver<T> resolver;
+   private final PermissionDynamic[] dynamics;
```
`PermissionDynamic<?>`

Also, why an array and not a `List` or `Collection`?

JTK222 replied:
In favor of being able to use varargs. I know that you have fancy alternatives for this in Kotlin.
But in Java this is the cleanest solution.

TheSilkMiner commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionNode.java:
```diff
@@ -0,0 +1,121 @@
+    * @param dynamics PermissionDynamic is a dynamic component for permission nodes, similar to BlockState Properties.
+    *                 They <strong>must</strong> be paassed into the constructor if you want to use them.
+    */
+   public PermissionNode(String nodeName, PermissionType<T> type, PermissionResolver<T> resolver, PermissionDynamic... dynamics)
```
Do you really want to expose the raw `nodeName` constructor?

TheSilkMiner commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionNode.java:
```diff
@@ -0,0 +1,121 @@
+      this.resolver = resolver;
+   }
+
+   /**
+    * <p>Allows you to set a human-readable name & description for your Permission.
+    * It is recommended to keep descriptions short, but there is no hard limit.</p>
+    *
+    * @param readableName a readable name for the PermissionNode
+    * @param description description for the PermissionNode
+    * @return itself with the new information set.
+    */
```
I'd add that permission Handlers may use this information in game for something. And also a suggested max length, like at most 200 chars or something.

Alpvax replied:
How could we adjust this to allow translations?
Do we want to change to TextComponents, or just set translation keys here instead?

TheSilkMiner commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionTypes.java:
```diff
@@ -17,19 +17,17 @@
-    ALL,
-    OP,
-    NONE
+   public static final PermissionType<Boolean> BOOL = new PermissionType<>(Boolean.class, "Boolean");
+   public static final PermissionType<Integer> INT = new PermissionType<>(Integer.class, "Integer");
+   public static final PermissionType<String> STRING = new PermissionType<>(String.class, "String");
+   public static final PermissionType<Component> COMPONENT = new PermissionType<>(Component.class, "Component");
```
Uppercase names? Also, some Javadoc on examples of usage might be neat. It's probably redundant for booleans, but components and strings would benefit IMO.

TheSilkMiner reviewed (changes requested):
Some clarification required on the documentation. Other than that, it looks fine.

Alpvax commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionNode.java:
```diff
@@ -0,0 +1,121 @@
+      return resolver;
+   }
+
+   public String getReadableName()
```
Do we want to add a way to return ITextComponents instead (e.g. for translation of this and description)
Or at least a way to have translated names and descriptions?

Alpvax commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionNode.java:
```diff
@@ -0,0 +1,121 @@
+      return readableName;
+   }
+
+   public String getDescription()
```
Make translatable? (see above)

Alpvax reviewed (commented):
How do we want to handle non-english permission names/descriptions?

JTK222 commented:
> How do we want to handle non-english permission names/descriptions?

Simple: We don't.
The Server doesn't have the ability to have multiple translations.
Forge patches it to load the English language file, but that's about it.
Changing the strings to Components would also make it harder for permission handlers,
to use those names & descriptions outside the game.
Which at least imo is not a good trade of for eventually sending the components to the client.
And finally, that would add a client side component, to what is supposed to be a server side api.

However, if you have a solution for the above problems, I am open to it

Alpvax commented:
> Simple: We don't.
> The Server doesn't have the ability to have multiple translations.
> Forge patches it to load the English language file, but that's about it.

That means that as long as there is an English translation, it will work on a server. Doesn't explicitly disallow alternative translations (on client)

> Changing the strings to Components would also make it harder for permission handlers,
> to use those names & descriptions outside the game.
> Which at least imo is not a good trade of for eventually sending the components to the client.

That is a very good reason not to have Components, agreed. The manager can always create them (potentially from a translation key) for the client if required.

> And finally, that would add a client side component, to what is supposed to be a server side api.

It is, but I could arguably see a manager adding a command-based (or other in-game) configuration system. Having localised names/descriptions on the client could be useful in that case.

> However, if you have a solution for the above problems, I am open to it

Excluding TextComponents, that leaves the old translate-a-string approach (which admittedly I don't know if it still exists): Use I18n to directly translate a string into another string. As long as the translation key is accessible, the manager can send them to the client to be localised.


I don't feel super strongly about it (I am an English speaker, so it actually doesn't affect me), but I don't want to accidentally exclude the rest of the world (resulting in this API still not seeing much use).

JTK222 commented:
I've added a way to Translation Permission Names & Descriptions.
It'll provide a LazyOptional with auto generated translation keys.
(The node name prefixed with: "permission.<name|desc>")

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/exceptions/UnregisteredPermissionException.java:
```diff
@@ -0,0 +1,9 @@
+    public UnregisteredPermissionException(PermissionNode node) {
+        super(String.format("Tried to query PermissionNode '%s' although it has not been Registered", node.getNodeName()));
+    }
+}
```
Formatting: braces on newlines.

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/exceptions/UnregisteredPermissionException.java:
```diff
@@ -0,0 +1,9 @@
+package net.minecraftforge.server.permission.exceptions;
```
Missing license header.

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/PermissionAPI.java:
```diff
@@ -19,80 +19,77 @@
-    {
-        return permissionHandler;
-    }
+   private PermissionAPI() {}
```
I'm not entirely sure if this is correct or incorrect in light of the formatting guide of braces on newlines. Best to be on the safe side and put them on newlines.

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/events/PermissionGatherEvent.java:
```diff
@@ -0,0 +1,94 @@
+import java.util.Set;
+
+/**
+ * <p>Fired during Server startup<br/>
```
The lead sentence of a javadoc should summarize what the class does or is used for. When a javadoc comment is processed into documentation HTML by `javadoc`, the lead sentence is used in class/method/field listings as a way of giving an at-a-glance overview of what the class does.

For example, a more informative lead sentence would be `Fired to gather information for the permissions API, such as the {@link PermissionHandler} and {@link PermissionNode}s.`

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/handler/IPermissionHandler.java:
```diff
@@ -0,0 +1,73 @@
+   /**
+    * {@return an identifier for the PermissionHandler, used for logging purposes}
+    */
+   ResourceLocation getPermissionHandlerName();
```
Does the method name need to be qualified with the `PermissionHandler` part? Given that the method exists in the `IPermissionHandler` class, and the javadoc already identifies it as being an identifier for the permission handler. 

Also, should it be `getName()` for the 'name', or `getIdentifier()` as it's an identifier?

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/handler/IPermissionHandler.java:
```diff
@@ -0,0 +1,73 @@
+   ResourceLocation getPermissionHandlerName();
+
+   /**
+    * @param node to be registered, should not be called by mods.
```
The part about this being a method not for usage by mods / anything other than the permission API should be clarified more. This should also link to the `PermissionGatherEvent`, to redirect mod developers on what API they must use.

Additionally, is it undefined behavior if the permission API somehow passes in the same node twice to this method? Is it guaranteed by the permission API that no two nodes passed to this method will be the same, for at least the lifetime of the server (or less)?

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/PermissionTest.java:
```diff
@@ -0,0 +1,87 @@
+package net.minecraftforge.debug;
```
Missing license header.

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionTypes.java:
```diff
@@ -0,0 +1,46 @@
+
+   @Nullable
+   public static PermissionType<?> getTypeByName(String name){
+      return switch(name){
```
Formatting: braces on newlines.

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionNode.java:
```diff
@@ -0,0 +1,162 @@
+   /**
+    * {@return A LazyOptional of the translation key, left is the Permission Name, right is the Description}
+    */
+   public LazyOptional<Pair<String, String>> getTranslationKeys(){
```
Formatting: braces on newlines.

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionDynamicContextKey.java:
```diff
@@ -0,0 +1,38 @@
+ */
+public record PermissionDynamicContextKey<T extends StringRepresentable> (Class<T> typeToken, String name)
+{
+   public PermissionDynamicContext<T> createContext(T value){
```
Formatting: braces on newlines.

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/handler/DefaultPermissionHandler.java:
```diff
@@ -0,0 +1,64 @@
+   private ResourceLocation name;
+   private Collection<PermissionNode<?>> registeredNodes;
+
+   public DefaultPermissionHandler(){
```
Formatting: braces on newlines.

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/PermissionAPI.java:
```diff
@@ -19,80 +19,77 @@
-    private static final Logger LOGGER = LogManager.getLogger();
-    
-    private static IPermissionHandler permissionHandler = DefaultPermissionHandler.INSTANCE;
+   private static Logger LOGGER = LogManager.getLogger();
```
`LOGGER` line should be `final` (no reason not to be).

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/PermissionAPI.java:
```diff
@@ -19,80 +19,77 @@
+      PermissionGatherEvent.Handler event = new PermissionGatherEvent.Handler(activeHandler);
+      MinecraftForge.EVENT_BUS.post(event);
+      IPermissionHandler newHandler = event.getCurrentHandler();
+      LOGGER.warn("Replacing {} with {}", activeHandler.getPermissionHandlerName(), newHandler.getPermissionHandlerName());
```
This should not be logging at the warning level, as that level is for "[a]n event that might possible [sic] lead to an error" (see its [javadocs](https://logging.apache.org/log4j/2.x/log4j-api/apidocs/org/apache/logging/log4j/Level.html#WARN)). At best, this should be at the info level.

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/handler/IPermissionHandler.java:
```diff
@@ -0,0 +1,73 @@
+    */
+   void registerNode(PermissionNode<?> node);
+
+   /**
+    * {@return a collection of all registered Nodes}
+    */
+   Collection<PermissionNode<?>> getRegisteredNodes();
```
Is this a modifiable or unmodifiable collection? Is this a snapshot of the registered nodes at the time the method was called, or is it a live view of all registered nodes?

JTK222 replied:
Would you say that specifying it as unmodifieable in javadoc is enough?
Or should I use the google.common (or whatever exactly the package is) ImmutableList?

TheSilkMiner replied:
`Collections.unmodifiableCollection(original)`?

sciwhiz12 replied:
Note that there's a difference between an immutable collection and an unmodifiable collection: the former means the list's contents will never change and cannot be changed through that object, while the latter only means the list's contents cannot be changed through that object -- it's possible for an unmodifiable collection to change its contents.

As @TheSilkMiner points out, using `Collections.unmodifiableCollection(Collection<T>)` creates an unmodifiable live view of the passed-in collection. So my recommendation is to mark the method as returning  `an unmodifiable view of the collection of registered permission nodes`, then return in the default permission handler a `Collections.unmodifiableCollection(registeredNodes)`. 

The return type shouldn't change (other than perhaps considering a move to `Set` instead of `Collection`, for the guarantee of there never being two `PermissionNode`s in the set which are `equals(Object)` with each other.

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/handler/IPermissionHandler.java:
```diff
@@ -0,0 +1,73 @@
+    * @param <T> type of the queried PermissionNode
+    * @return a value of type {@code <T>}, that the combination of Player and PermissionNode map to.
+    */
+   <T> T getPermissions(ServerPlayer player, PermissionNode<T> node, PermissionDynamicContext<? extends StringRepresentable>... context);
```
I'd suggest losing the `s`: `getPermission`, since there's only one permission node and corresponding value here.

sciwhiz12 reviewed (changes requested):
Some notes:
* Indentation is 4 spaces (no tabs) in Forge classes. The exception is for Minecraft classes, which must use 3 spaces (to follow the output of the decompiled source).
* The first paragraph in a javadoc comment does not need to be surrounded by `<p>` tags, only the second and further paragraphs need them. Additionally, I suggest putting a blank line between each paragraph, for readability when looking at the javadocs in source form.

Alpvax commented:
> I've added a way to Translation Permission Names & Descriptions.
> It'll provide a LazyOptional with auto generated translation keys.
> (The node name prefixed with: "permission.<name|desc>")

Just looked at how you've implemented it. It means that any mod adding translations will have to duplicate the name and description in code and the en_us language file. I imagine that will discourage people from adding the translations.

I would suggest moving all information to the language file, and translating it on the server as well.
If the friendly name and description are optional, you could always have a bool for each one.

Currently, if addInformation is not called, the translation optional is not initialised, which could cause errors down the line.

Alpvax commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionNode.java:
```diff
@@ -0,0 +1,162 @@
+    * Calls {@link PermissionNode#PermissionNode(String, PermissionType, PermissionResolver, PermissionDynamicContextKey[])}
+    * with "modname.nodename" as the first parameter
+    */
+   public PermissionNode(String modName, String nodeName, PermissionType<T> type, PermissionResolver<T> resolver, PermissionDynamicContextKey... dynamics)
```
Could we change `modName` to `modId` to be more consistent with everywhere else?

Alpvax commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionNode.java:
```diff
@@ -0,0 +1,179 @@
+
+   /**
+    * Calls {@link PermissionNode#PermissionNode(String, PermissionType, PermissionResolver, PermissionDynamicContextKey[])}
+    * with "modname.nodename" as the first parameter
```
Missed changing modName to modId in the javadoc

JTK222 commented:
> Just looked at how you've implemented it. It means that any mod adding translations will have to duplicate the name and description in code and the en_us language file. I imagine that will discourage people from adding the translations.

Just checked it, as it seems you'd have redundancy anyway.
Server side translations have to be in the data folder (for mods, forge and vanilla are special cased).
Additionally the question is whether the code for loading those language files will stay or whether that feature will be removed?

> Currently, if addInformation is not called, the translation optional is not initialised, which could cause errors down the line.

I fixed that, with some error handling, but might be changed if we can really 100% move this to server translations

From the Code for the Server Side translation class:
```java
//TODO, this should be re-evaluated now that ITextComponents are passed everywhere instaed of strings.
public class ForgeI18n {
```

Edit: After some convo on the Discord we ended up with the conclusion that we can move to Components for now.
The server can parse TranslatableComponents, but requires an en_us language file in the data folder of mods.
If this needs to be changed at one point, special handling can be added.

Alpvax commented:
> Just checked it, as it seems you'd have redundancy anyway.
> Server side translations have to be in the data folder (for mods, forge and vanilla are special cased).

I didn't realise that was the case. Do data generators handle that automatically? If so that would help.

> Edit: After some convo on the Discord we ended up with the conclusion that we can move to Components for now.
> The server can parse TranslatableComponents, but requires an en_us language file in the data folder of mods.
> If this needs to be changed at one point, special handling can be added.

Sounds good to me.

JTK222 commented:
> I didn't realise that was the case. Do data generators handle that automatically? If so that would help.

Doubt it, server side translations are a rarely used feature.
But if it really bothers people, there are many ways to fix it, for example a gradle script that copies the lang file, or merges a client and server lang file.

JTK222 commented:
I want to point out, the build failed because of concurrency issues on the CLI.
It builds locally and the PR is so far finished if there is no more criticism.
Sadly, I don't have anything for another commit to trigger the CLI again due to that.

Alpvax commented:
> Sadly, I don't have anything for another commit to trigger the CLI again due to that.

Does force pushing work?

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/events/PermissionGatherEvent.java:
```diff
@@ -0,0 +1,105 @@
+                if (this.nodes.contains(node))
+                    throw new IllegalArgumentException("Tried to register duplicate PermissionNode '" + node.getNodeName() + "'");
+                this.nodes.add(node);
+            }
```
a) Please put the brackets on newlines.
b) The two calls to `contains` then `add` can be replaced with a single `add` in the `if`-condition:
> Adds the specified element to this set if it is not already present (optional operation). More formally, adds the specified element `e` to this set if the set contains no element `e2` such that `Objects.equals(e, e2)`. If this set already contains the element, the call leaves the set unchanged and returns `false`.
> ~ *[`Set#add(E)`](https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/util/Set.html#add(E)), Java SE 16 & JDK 16*

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/exceptions/UnregisteredPermissionException.java:
```diff
@@ -17,19 +17,14 @@
-public enum DefaultPermissionLevel
+import net.minecraftforge.server.permission.nodes.PermissionNode;
+
+public class UnregisteredPermissionException extends RuntimeException
```
In my opinion, this exception class should either be moved as a (package-)private static exception class instead of being part of the public API, or expanded to include a field with the `PermissionNode` that was attempted to be queried if it is to be part of the public API.

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionDynamicContext.java:
```diff
@@ -0,0 +1,51 @@
+    {
+        this.dynamic = dynamic;
+        this.value = value;
+    }
```
Given that `PermissionDynamicContextKey#createContext` exists which calls this constructor, should this constructor remain public or be moved to package-private visibility?

JTK222 replied:
Good Point, changed to be package private

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionNode.java:
```diff
@@ -0,0 +1,174 @@
+     */
+    public PermissionNode(ResourceLocation nodeName, PermissionType<T> type, PermissionResolver<T> resolver, PermissionDynamicContextKey... dynamics)
+    {
+        this(nodeName.getNamespace() + "." + nodeName.getPath(), type, resolver, dynamics);
```
A bit nitpicky, but this could probably forward to the next constructor instead (`PermissionNode(String, String, PermissionType<T>, PermissionResolver<T>, PermissionDynamicContextKey`).

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionNode.java:
```diff
@@ -0,0 +1,174 @@
+    public PermissionNode setInformation(@Nonnull Component readableName, @Nonnull Component description)
+    {
+        Preconditions.checkNotNull(readableName, "Readable name for PermissionNodes must not be null {}", this.nodeName);
+        Preconditions.checkNotNull(description, "Description for PermissionNodes must not be null {}", this.nodeName);
```
Checking the javadocs for [`Preconditions.checkNotNull`](https://guava.dev/releases/21.0/api/docs/com/google/common/base/Preconditions.html#checkNotNull-T-java.lang.String-java.lang.Object...-) reveals that the placeholder should be `%s` (`{}` is used by Log4j). Additionally, you can reformat the message to "[Readable name/Description] for %s must not be null" to be more readable.

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionNode.java:
```diff
@@ -0,0 +1,174 @@
+ * There are no restrictions on their amount. It is also not mandatory to provide a value for all dynamics in a permission query.
+ * See the above link for more information.</p>
+ *
+ * <p>Each node should be registered via the {@link PermissionGatherEvent.Nodes} and stored statically in a field.
+ * That instance should then be reused every-time a permission check needs to be performed via {@link net.minecraftforge.server.permission.PermissionAPI#getPermissions(ServerPlayer, PermissionNode, PermissionDynamicContext[])}.</p>
```
I almost wonder if there should be an equivalent to `DeferredRegister` and `RegistryObject` for permission nodes here.

JTK222 replied:
It might make sense, but as PermissionNodes cannot be overridden, I don't see a major benefit for this.
It would only cause overhead as of right now.

sciwhiz12 replied:
Fair enough, though my thought is it would allow defining the permission node in a `static final` field and having it handle automatically registering the permission nodes through the event, but that is more for modder's convenience's sake which they could do on their side too.

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/PermissionTest.java:
```diff
@@ -0,0 +1,106 @@
+{
+   private static final boolean isEnabled = true;
+   private static final Logger LOGGER = LogManager.getLogger();
+
```
Please fix the formatting of this class to use 4 spaces, not 3 spaces.
Also, it may be worth considering changing the modid to `permission_test`, but that's up to you.

JTK222 replied:
I assume the 4 spaces are to match the remaining tests?
Wouldn't it be better to remain with the style guide?

sciwhiz12 replied:
To my knowledge, all non-Minecraft classes (that is, classes which aren't used/touched by the patcher) uses 4 spaces for indentation. 3 spaces only exist in Minecraft classes due to that being the output of Fernflower.

JTK222 replied:
Of course it has to be complex enough to cause trouble with auto formatting...
Any ways, fixed that for now.

sciwhiz12 commented on patches/minecraft/net/minecraft/client/server/IntegratedServer.java.patch:
```diff
@@ -1,13 +1,15 @@
        this.m_129989_(this.m_129791_() + " - " + this.m_129910_().m_5462_());
 -      return true;
++      net.minecraftforge.server.permission.PermissionAPI.gatherPermissionHandler();
++      net.minecraftforge.server.permission.PermissionAPI.gatherPermissionNodes();
```
Could this perhaps be moved into the hook for `FMLServerStartingEvent` instead?

sciwhiz12 commented on patches/minecraft/net/minecraft/server/dedicated/DedicatedServer.java.patch:
```diff
@@ -33,6 +33,8 @@
  
 -         return true;
++         net.minecraftforge.server.permission.PermissionAPI.gatherPermissionHandler();
++         net.minecraftforge.server.permission.PermissionAPI.gatherPermissionNodes();
```
Same here: could this perhaps be moved into the hook for `FMLServerStartingEvent` instead?

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/PermissionAPI.java:
```diff
@@ -19,80 +19,79 @@
+        PermissionGatherEvent.Nodes event = new PermissionGatherEvent.Nodes();
+        MinecraftForge.EVENT_BUS.post(event);
+
+        for (PermissionNode<?> node : event.getNodes()) {
```
Braces on newlines, please.

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/PermissionAPI.java:
```diff
@@ -19,80 +19,79 @@
-        // TODO Loader states Preconditions.checkState(Loader.instance().getLoaderState().ordinal() > LoaderState.PREINITIALIZATION.ordinal(), "Can't register permission nodes before Init!");
-        permissionHandler.registerNode(node, level, desc);
-        return node;
+        if (!activeHandler.getRegisteredNodes().contains(node)) throw new UnregisteredPermissionException(node);
```
This should be documented in the javadoc through a `@throws` tag.

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/events/PermissionGatherEvent.java:
```diff
@@ -0,0 +1,106 @@
+            {
+                if (!this.nodes.add(node))
+                    throw new IllegalArgumentException("Tried to register duplicate PermissionNode '" + node.getNodeName() + "'");
+                ;
```
Please remove this stray semicolon.

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionDynamicContext.java:
```diff
@@ -0,0 +1,51 @@
+    private PermissionDynamicContextKey<T> dynamic;
+    private T value;
+
+    PermissionDynamicContext(PermissionDynamicContextKey<T> dynamic, T value)
```
Since this class cannot become a record in order to preserve the package-private access level of its constructor, it may be worth adding a comment to that effect, but up to you.

While on the topic of records: since this class is effectively a record, it might be worth implementing `hashCode` and `equals` here.

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionType.java:
```diff
@@ -17,10 +17,11 @@
-
-import javax.annotation.ParametersAreNonnullByDefault;
+/**
+ * <p>Type of a Permission, use the existing Types in {@link PermissionTypes}</p>
```
What happens if a mod creates a `PermissionType` of their own, that is not defined in `PermissionTypes`? How would other mods or permission handlers react in such a situation?

sciwhiz12 commented on src/test/resources/META-INF/mods.toml:
```diff
@@ -168,5 +168,7 @@ license="LGPL v2.1"
 [[mods]]
     modId="full_pots_accessor_demo"
+[[mods]]
+    modId="permissiontest"
```
The `@Mod` class has its modid as `permission_test`, please correct this to match.

TheCurle replied:
This still has not been fixed. 
```
@Mod("permission_test")
```

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionNode.java:
```diff
@@ -0,0 +1,174 @@
+
+    /**
+     * @param nodeName The identifier of a node, recommended identifier structure: "modid.path.for.node"
+     * @param type     type object for the PermissionNode, only use types in {@link PermissionTypes}, if you need more PR them
```
Remove the latter phrase "if you need more PR them", as that's already mentioned in the javadocs for `{@link PermissionTypes}`.

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionNode.java:
```diff
@@ -0,0 +1,174 @@
+     * @param type     type object for the PermissionNode, only use types in {@link PermissionTypes}, if you need more PR them
+     * @param resolver Default resolver for the permission, can but doesn't have to be used to by PermissionHandlers
+     * @param dynamics PermissionDynamicContextKey is a dynamic component for permission nodes, similar to BlockState Properties.
+     *                 They <strong>must</strong> be paassed into the constructor if you want to use them.
```
Nitpick'd typo:
```suggestion
     *                 They <strong>must</strong> be passed into the constructor if you want to use them.
```

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionNode.java:
```diff
@@ -0,0 +1,174 @@
+    /**
+     * @param nodeName The identifier of a node, recommended identifier structure: "modid.path.for.node"
+     * @param type     type object for the PermissionNode, only use types in {@link PermissionTypes}, if you need more PR them
+     * @param resolver Default resolver for the permission, can but doesn't have to be used to by PermissionHandlers
```
If this is the default resolver for the permission, the parameter name and field name should also be named to communicate that: `defaultResolver`.

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionNode.java:
```diff
@@ -0,0 +1,174 @@
+        return description;
+    }
+
+    public interface PermissionResolver<T>
```
This needs javadocs as its part of the public modder-used API.

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionTypes.java:
```diff
@@ -0,0 +1,48 @@
+/**
+ * Default PermissionTypes, if you need additional ones, please PR it.
+ */
+public final class PermissionTypes
```
Since this is a `final` utility class, should also have a `private` constructor to match.

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/PermissionTest.java:
```diff
@@ -0,0 +1,107 @@
+@Mod("permission_test")
+public class PermissionTest
+{
+    private static final boolean isEnabled = true;
```
A bit nitpicky, but the conventional name for this field is `ENABLED`.

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionTypes.java:
```diff
@@ -0,0 +1,48 @@
+public final class PermissionTypes
+{
+    public static final PermissionType<Boolean> BOOL = new PermissionType<>(Boolean.class, "boolean");
+    public static final PermissionType<Integer> INT = new PermissionType<>(Integer.class, "integer");
```
Was there a particular reason why these were named `BOOL` and `INT` and not `BOOLEAN` and `INTEGER`?

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/PermissionTest.java:
```diff
@@ -0,0 +1,107 @@
+            LOGGER.error(e.getMessage());
+            return false;
+        }
+    }
```
What does this do exactly, for the record? And would it be worth exposing this as part of the public API, for modder's convenience?

JTK222 replied:
All it does is catch the UnregisteredPermissionException, as it's otherwise swallowed by other resulting exceptions in the command tree. 

Modders shouldn't really run into this issue when using the API correctly (not forgetting to register their perms).
And they should already have some proper error handling for everything that might fail in their commands.

The method is only in the test for convenience and as a failing test case is required.

sciwhiz12 replied:
Sounds fine to me, though that should be explained in a comment (the javadoc).

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/events/PermissionGatherEvent.java:
```diff
@@ -0,0 +1,105 @@
+
+        public Nodes()
+        {
+            this.nodes = new HashSet<>();
```
Why is the field initialized in the constructor? This could be moved to be part of the field definition, and marked `final` as well.

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/events/PermissionGatherEvent.java:
```diff
@@ -0,0 +1,105 @@
+
+        public Collection<PermissionNode<?>> getNodes()
+        {
+            return nodes;
```
Should we perhaps consider wrapping the returned collection via [`Collections#unmodifiableCollection(Collection)`](https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/util/Collections.html#unmodifiableCollection(java.util.Collection))?

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/handler/DefaultPermissionHandler.java:
```diff
@@ -0,0 +1,66 @@
+
+    public DefaultPermissionHandler()
+    {
+        this.name = new ResourceLocation("forge", "default_handler");
+        this.registeredNodes = new HashSet<>();
```
Again, why not have this be part of the field declarations and marked `final`?

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/handler/IPermissionHandler.java:
```diff
@@ -0,0 +1,77 @@
+public interface IPermissionHandler
+{
+    /**
+     * {@return an identifier for the PermissionHandler, used for logging purposes}
```
We should leave out the `used for logging purposes`, to keep open the possibility of having other uses for the handler identifier (even if there is none other than logging currently).

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/handler/IPermissionHandler.java:
```diff
@@ -0,0 +1,77 @@
+     * @param <T>     type of the queried PermissionNode
+     * @return a value of type {@code <T>}, that the combination of Player and PermissionNode map to.
+     */
+    <T> T getPermission(ServerPlayer player, PermissionNode<T> node, PermissionDynamicContext<? extends StringRepresentable>... context);
```
Is the `? extends StringRepresentable` needed here? Wouldn't the generic type in the `PermissionDynamicContext` class of `T extends StringRepresentable` already force the generic type to always be inheriting from `StringRepresentable`?

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionNode.java:
```diff
@@ -0,0 +1,180 @@
+    private final PermissionResolver<T> defaultResolver;
+    private final PermissionDynamicContextKey<?>[] dynamics;
+
+    private Component readableName;
+    private Component description;
```
As these are `null` by default, these should be marked `@Nullable` along with their getters.

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionNode.java:
```diff
@@ -0,0 +1,180 @@
+    public boolean equals(Object o)
+    {
+        if (this == o) return true;
+        if (!(o instanceof PermissionNode)) return false;
+        PermissionNode<?> that = (PermissionNode<?>) o;
```
Pattern matching `instanceof` can be used here:
```suggestion
        if (!(o instanceof PermissionNode that)) return false;
```

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionDynamicContext.java:
```diff
@@ -0,0 +1,70 @@
+    public boolean equals(Object o)
+    {
+        if (this == o) return true;
+        if (!(o instanceof PermissionDynamicContext)) return false;
+        PermissionDynamicContext<?> that = (PermissionDynamicContext<?>) o;
```
Pattern matching `instanceof` can be used here:
```suggestion
        if (!(o instanceof PermissionDynamicContext that)) return false;
```

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionType.java:
```diff
@@ -0,0 +1,74 @@
+    public boolean equals(Object obj)
+    {
+        if (obj == this) return true;
+        if (obj == null || obj.getClass() != this.getClass()) return false;
+        var that = (PermissionType) obj;
```
As this class is `final`, we know that no other classes can extend this class, so a pattern matching `instanceof` is sufficient here:
```suggestion
        if (!(obj instanceof PermissionType that)) return false;
```

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/nodes/PermissionType.java:
```diff
@@ -0,0 +1,74 @@
+        if (obj == this) return true;
+        if (obj == null || obj.getClass() != this.getClass()) return false;
+        var that = (PermissionType) obj;
+        return Objects.equals(this.typeToken, that.typeToken) &&
+            Objects.equals(this.typeName, that.typeName);
```
Just as a note, `Object#equals(Object)` could also have been used here since we know these fields are always non-`null`. This as-is is fine though.

sciwhiz12 commented on src/main/java/net/minecraftforge/server/permission/package-info.java:
```diff
@@ -1,26 +0,0 @@
- * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
- */
-
-@ParametersAreNonnullByDefault
-@MethodsReturnNonnullByDefault
-package net.minecraftforge.server.permission;
```
Was there a particular reason why the `package-info`s were deleted?

sciwhiz12 reviewed (changes requested):
🤞 After this round of review comments, I think this PR is ready for assignment and merging.

JTK222 committed:
Wipe of old PermissionAPI


New PermissionAPI Implementation


Added Patch, removed unused Imports


Added Copyright Notice


Fixed missing vararg for IPermissionHandler


Added Test Mod


Added Javadoc


Added Information to PermissionNodes

Information consists of a human readable name and a description, both are optional. An example use case might be producing a file containing all available permissions.

Also fixed unused imports in PermissionTest
Added Permission Setup to integrated Server


Fixed Javadoc, improved test mod


Improved Java Doc, added translation ability for perm descriptions.


Made most of the suggested changes


Fixed Formatting


Changed getRegisteredNodes method


Added ImplNote to PermissionGatherEvent.Nodes#addNodes


Added Duplicate PermissionNodeHandling


Removed Impl Note as it's invalid now


Improved Javadoc


Fixed unused imports (again)


Added Error Handling for PermissionNode#setInformation


Fixed initialization of LazyOptional in PermissionNode


Changed ReadableName and Description to Components

Allows for translations, but requires an english language file in the data directory.
Moved Event Hooks to Server Lifecycle hook, removed patches


Fixed Formatting, Access levels and expanded exception


Fixed Patches


Should finally fix all incorrect indents


More smaller fixes


Improved explanation for the Test Mod


Final Fixes

Re-added package-info
Made changes requested by schiwiz
Added Ability to check Permissions of Offline Players


Improved PermissionResolver for offline players


Removed need for StringRepresentable in PermissionDynamicContext


Added Nullable Annotation to fields in PermissionNode


Fixed fups during rebase

JTK222 committed:
Re-Introduced PermissionAPI in lifecycle hooks

LexManos commented on src/main/java/net/minecraftforge/server/permission/PermissionAPI.java:
```diff
@@ -19,80 +19,101 @@
-        Preconditions.checkArgument(!node.isEmpty(), "Permission node can't be empty!");
-        return permissionHandler.hasPermission(profile, node, context);
+        PermissionGatherEvent.Handler event = new PermissionGatherEvent.Handler(activeHandler);
+        MinecraftForge.EVENT_BUS.post(event);
```
This needs a user config to determine the winner. 
And they need to be registered by name.
So that if two mods add handlers, the user can be like 'screw you I want forge:default'

LexManos commented on src/main/java/net/minecraftforge/server/permission/handler/DefaultPermissionHandler.java:
```diff
@@ -0,0 +1,72 @@
+    @Override
+    public Set<PermissionNode<?>> getRegisteredNodes()
+    {
+        return Collections.unmodifiableSet(this.registeredNodes);
```
Views like this should be cached

LexManos commented on src/main/java/net/minecraftforge/server/permission/handler/IPermissionHandler.java:
```diff
@@ -0,0 +1,90 @@
+     *
+     * @param node PermissionNode to be registered
+     */
+    void registerNode(PermissionNode<?> node);
```
Should this instead be a factory method so that people cant some in at runtime and add random nodes?

JTK222 replied:
I think that's unnecessary, the final active PermissionHandler isn't accessible.
So registering new nodes outside the events isn't really possible.
Additionally, this would require mods to reconstruct their PermissionNodes whenever a singleplayer world is loaded.
Which I fear might tempt people to return to the old way of creating dynamic permissions, 
rather than using the new more universal approach.

JTK222 committed:
Added config option to choose PermissionHandler

JTK222 committed:
Default PermissionHandler now caches immutable set of registered nodes.

JTK222 committed:
Added a method to check which PermissionHandler is active

JTK222 committed:
Removed unused Import in ForgeConfig

LexManos committed:
Properly cache the view without threading issues

https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#unmodifiableSet-java.util.Set-
The views are passthrough, there is no need to keep recreating them.
In fact doing so exposes a timeframe where the return is null and can cause a NPE in consumers.

LexManos commented:
Look, I personally dislike the entire concept of this PR. I think the new design is overly complex and exposes many many potential avenues for things to screw up and modders to do things incorrectly. While also not adequately explaining what is wrong fundamentally with the old system. Beyond the fact that modders are doing what they do, and using it wrong. However, there are others on the team who like this concept and want to see it implemented. So I am purely approaching this now from a functional/security concept. As such i've discussed several issues internally with other members of the team and they will be forwarding them when they gets a chance. However, {this is mainly a note for the triage team}, from this point on I'll be assigning this to Curle, as she can pull it when she and the others deem it ready.

JTK222 committed:
Reworked PermissionAPI initialization Process

JTK222 committed:
Added missing Header

JTK222 committed:
Fixed Formatting

JTK222 committed:
Fixed Testmod modid

JTK222 committed:
Made more changes as discussed in Discord

JTK222 committed:
Final Changes

TheCurle commented:
After a lot of back-and-forth in the Discord, i feel like this has reached a point that the majority of us are happy with.

All of the concerns raised so far have been mitigated rather well, especially concerning the global state and startup hooks.

It meets our criteria for being a great addition to the API, but remember that this should still be seen as experimental. Things will change and improve over time as we understand how people use the API, and for what.

In that sense, it's a great start and hopefully will lead to lots of great mod features implemented in the future.

maxanier commented:
I like that the PermissionAPI got some love, but I am missing some documentation on the `PermissionDynamicContext`.
I think I got it to work, but some things feel akward, so maybe I didn't understood it correctly. I haven't found any usage of it in the test mod or on here. Also I found the quote "Small Explanation: PermissionDynamics need to be registered together with a PermissionNode." here, but I can find any way do that.

1. When simply trying to get a permission without a context, IntelliJ warns about `Unchecked generics array creation for varargs parameter` which is annoying. Of course it can be suppressed, but I do not want to this globally, so I have to add a suppress comment every time

2. I can create a `PermissionDynamicContextKey` just fine for any generic type I like, but when I want to create a `PermissionDynamicContext` of it and pass it to `getPermission` , the generic type must extend `StringRepresentable`. This means I can't use a String, but must have a class that extends `StringRepresentable`. However, I still need to provide a object-to-string mapper, even though the object is practically required to be a `StringRepresentable`.
```java
    public static PermissionDynamicContextKey<IAction> ACTION_CONTEXT = new PermissionDynamicContextKey<>(IAction.class, "action", StringRepresentable::getSerializedName);
...
    private boolean isActionAllowedPermission(IAction<T> action) {
        if(player.getRepresentingPlayer() instanceof ServerPlayer serverPlayer){
            //noinspection unchecked
            return PermissionAPI.getPermission(serverPlayer, Permissions.ACTION, Permissions.ACTION_CONTEXT.createContext(action));
        }
        return true;
    }
```

Is this correct or did I miss something?

JTK222 commented:
Seems like this was missed in the PR, 
it should not have to extend StringRepresentable I'll attempt to fix this quickly.

Beyond that on a first glance your code snippet seems fine.
(However, your IAction seems to be a generic type, that's likely what causes the unchecked warning)
In regards to the missing documentation, feel free to message me on discord, I can explain some things there.
My name is: JTK222 | Lukas#7285

maxanier commented:
Thank you! I might message you if I dig deeper into the permission system, but for now everything seems to work.
Maybe one or two lines about the DynamicContext, or an addition to the test mod, would make it easier for other modders as well to understand how to use it, although it is not super complicated.

(I also get the unchecked generic array warning with the Boolean type as well, it just refers to the variable generic array of the context. But maybe I can just disable that specific warning globally as I have never had any issues with something like that anyway)
